examples/DataApi-Water-And-Energy-Usage/template/modules/l0/src/main/scala/com/my/currency/l0/custom_routes/CustomRoutes.scala

```diff
@@ -0,0 +1,154 @@
+import cats.syntax.applicativeError.catsSyntaxApplicativeErrorId
+import cats.syntax.flatMap.toFlatMapOps
+import cats.syntax.functor.toFunctorOps
+import com.my.currency.shared_data.calculated_state.CalculatedStateService
+import com.my.currency.shared_data.deserializers.Deserializers
+import com.my.currency.shared_data.types.Types.{AddressTransactionsWithLastRef, TxnRef, UpdateUsageTransaction}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import eu.timepit.refined.auto._
+import org.http4s._
+import org.http4s.circe.CirceEntityCodec.circeEntityEncoder
+import org.http4s.dsl.Http4sDsl
+import org.http4s.server.middleware.CORS
+import org.tessellation.currency.dataApplication.L0NodeContext
+import org.tessellation.ext.http4s.AddressVar
+import org.tessellation.routes.internal.{InternalUrlPrefix, PublicRoutes}
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.schema.address.Address
+
+case class CustomRoutes[F[_] : Async](
+  calculatedStateService: CalculatedStateService[F],
+  context               : L0NodeContext[F]
+) extends Http4sDsl[F] with PublicRoutes[F] {
+
+  @derive(encoder, decoder)
+  case class TransactionResponse(
+    transactionType   : String,
+    energyUpdateAmount: Long,
+    waterUpdateAmount : Long,
+    txnSnapshotOrdinal: SnapshotOrdinal,
+    txnHash           : String,
+    lastRef           : TxnRef
+  )
+
+  private object TransactionResponse {
+    def make(updateUsageTransaction: UpdateUsageTransaction, snapshotOrdinal: SnapshotOrdinal, txnRef: String): TransactionResponse = {
+      TransactionResponse(
+        updateUsageTransaction.transactionType,
+        updateUsageTransaction.energyUpdateAmount,
+        updateUsageTransaction.waterUpdateAmount,
+        snapshotOrdinal,
+        txnRef,
+        TxnRef(updateUsageTransaction.lastTxnOrdinal, updateUsageTransaction.lastTxnHash)
+      )
+    }
+  }
+
+  private def getAddressTransactionsFromState(
+    ordinal: SnapshotOrdinal,
+    address: Address
+  ): F[AddressTransactionsWithLastRef] = {
+    context.getCurrencySnapshot(ordinal).flatMap {
+      case None => new Exception(s"Could not fetch snapshot: ${ordinal.value.value}").raiseError[F, AddressTransactionsWithLastRef]
+      case Some(value) =>
+        value.dataApplication
+          .fold(
+            AddressTransactionsWithLastRef(TxnRef.empty, List.empty[UpdateUsageTransaction]).pure
+          )(value =>
+            Deserializers.deserializeState(value.onChainState) match {
+              case Left(_) => new Exception(s"Could not deserialize state of snapshot: ${ordinal.value.value}").raiseError[F, AddressTransactionsWithLastRef]
+              case Right(value) =>
+                Async[F].delay {
+                  val updateUsageTransactions =  value.updates.filter { t => t.owner == address }.sortBy(_.lastTxnOrdinal)(Ordering[SnapshotOrdinal].reverse)
+                  val referenceTransaction = updateUsageTransactions.head
```

Comment:
Good catch, I've added a `handleErrorWith` on  `getAllAddressTransactions` to raise the error with a custom message

Reply:
@IPadawans I don't think that's the correct way to go. Assuming it's ok to return the "empty" `AddressTransactionsWithLastRef` when `updateUsageTransactions` is empty, I suggest a refactor similar to this:

```
  private def getAddressTransactionsFromState(
    ordinal: SnapshotOrdinal,
    address: Address
  ): F[AddressTransactionsWithLastRef] = {
    val dataApplicationPart: F[Option[currency.DataApplicationPart]] =
      OptionT(context.getCurrencySnapshot(ordinal))
        .getOrRaise(new Exception(s"Could not fetch snapshot: ${ordinal.show}, ${address.show}"))
        .map(_.dataApplication)

    OptionT(dataApplicationPart)
      .semiflatMap(da => MonadThrow[F].fromEither(Deserializers.deserializeState(da.onChainState)))
      .map(_.updates.filter(_.owner === address).sortBy(_.lastTxnOrdinal)(Ordering[SnapshotOrdinal].reverse))
      .mapFilter(txs => txs.headOption.map(t => AddressTransactionsWithLastRef(TxnRef(t.lastTxnOrdinal, t.lastTxnHash), txs)))
      .getOrElse(AddressTransactionsWithLastRef(TxnRef.empty, List.empty[UpdateUsageTransaction]))
  }
```
Not sure if we can clean it up further to avoid creating two `OptionT` objects but verify that this is functionally equivalent to what you had.